File: plugins/collector/event/apps/database/memcache.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"strings"
)

type Memcache struct {
	version string
}

func (Memcache) Name() string { return "memcache" }

func (Memcache) Type() string { return "database" }

func (m Memcache) Version() string { return m.version }

func (m *Memcache) Match(p *process.Process) bool { return p.Name == "memcached" }

func (m *Memcache) Run(p *process.Process) (mapping map[string]string, err error) {
	result, err := apps.Execute(p, "--version")
	if err != nil {
		return nil, err
	}
	m.version = strings.Trim(strings.TrimPrefix(result, "memcached "), "\n")
	return
}

func init() {
	apps.Regist(&Memcache{})
}

File: plugins/collector/src/event/apps/database/memcache.rs
use crate::event::apps::{execute, AppProc, IApp};
use std::collections::HashMap;

pub struct Memcache {
    version: String,
}

impl Memcache {
    pub fn new() -> Self { Memcache { version: String::new() } }
}

impl IApp for Memcache {
    fn name(&self)     -> &'static str { "memcache" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "memcached" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "memcached 1.6.17"
        self.version = out.trim_start_matches("memcached ").trim().to_owned();
        Ok(HashMap::new())
    }
}
